### ISSUE TITLE
Tasklist for updating and documenting sample applications

### ISSUE THREAD

**Issue by henri-hulski** [morepath/morepath#434]
Here I create a task-list for issues #419, #411, #286 and #376.

We want to:

* remove autosetup from demo code (#419)
* update the morepath example applications to follow cookiecutter-style setup (#411)
* add tests to the samples and integrate them with travis and tox
* make a release after adding CHANGES.rst and changing setup.py to add a long description
  by using io.open with `encoding='utf-8'` (#286)
* document sample applications (#376)

Here a list with sample applications and what has be done:

* [morepath_cerebral_todomvc](https://github.com/morepath/morepath_cerebral_todomvc)
  - [x] remove autosetup
  - [x] update to cookiecutter-style
  - [x] add tests
  - [x] make a release
  - [x] document sample
* [morepath_reactredux](https://github.com/morepath/morepath_reactredux)
  - [x] remove autosetup
  - [x] update to cookiecutter-style
  - [x] add tests
  - [x] make a release
  - [x] document sample
* [morepath_batching](https://github.com/morepath/morepath_batching)
  - [x] remove autosetup
  - [x] update to cookiecutter-style
  - [x] add tests
  - [x] make a release
  - [x] document sample
* [morepath_sqlalchemy](https://github.com/morepath/morepath_sqlalchemy)
  - [x] remove autosetup
  - [x] update to cookiecutter-style
  - [x] add tests
  - [x] make a release
  - [x] document sample
* [morepath_rest_dump_load](https://github.com/morepath/morepath_rest_dump_load)
  - [x] remove autosetup
  - [x] update to cookiecutter-style
  - [x] add tests
  - [x] make a release
  - [x] document sample
* [morepath_static](https://github.com/morepath/morepath_static)
  - [x] remove autosetup
  - [x] update to cookiecutter-style
  - [x] add tests
  - [x] make a release
  - [x] document sample
* [morepath_wiki](https://github.com/morepath/morepath_wiki)
  - [x] remove autosetup
  - [x] update to cookiecutter-style
  - [x] add tests
  - [x] make a release
  - [x] document sample

**Comment by henri-hulski** [comment]
I think it would be a good idea to add a CHANGES.rst and create an initial release on Github for the sample apps with the release date to the last commit.
Like this we can track the changes better.
After updating and providing the changes we can make a second release.

**Comment by henri-hulski** [comment]
@faassen 
When you decide which samples you want to keep I can just add CHANGES.rst, update the setup.py and make initial releases. (If you not prefer to do it yourself.)

After we can go on.
Replacing autosetup with autoscan can be done while updating to cookiecutter-style. I can take also a look at this.
Maybe it would also be nice to add some tests where missing.

I think for writing documentation I'm not the best choice. Maybe someone else.

**Comment by faassen** [comment]
* ``morepath_reactredux`` I'd like to keep; it's relatively new anyway and has some real moving parts. **keep**

* ``morepath_hello`` is more or less the quickstart example. It has an out of date autocommit in there. I'm not entirely sure we should get rid of it -- it's a nice way to get the thing installed and working, but then again the cookiecutter is a more official way. So I am leaning towards getting rid of it. **remove**

* ``morepath_batching`` I'd like to keep. It's quite involved and demonstrates a lot of morepath principles along with templating. it's referenced from here http://blog.startifact.com/posts/morepath-batching-example.html  **keep**

* while ``morepath_react`` demonstrates a relational backend and a react frontend, its react frontend is outdated a lot and rather peculiar to my way of thinking at the time. I think we should retire it, even though it contains quite a bit of code. **remove**

* ``morepath_sqlalchemy`` demonstrates integration with SQLAlchemy. I think that's useful. **keep**

* ``morepath_rest`` It demonstrates ``dump_json`` and such which is useful. It needs a readme that points to this blog entry http://blog.startifact.com/posts/better-rest-with-morepath-08.html  I think we should possibly rename it, as you can do REST fine without ``dump_json``. **keep**, with README and better name, perhaps ``morepath_dump_load`` or something.

*  ``morepath_static`` demonstrates bowerstatic integration. **keep*. But I think we should move the ``more.static`` documentation out of Morepath's main documentation and instead document it as part of ``more.static`` in the README and refer to it. This doc should point to this working example. New issue: #436

* ``morepath_demo`` is not really doing much useful anymore; its demo of Morepath purposes are better served by a more involved app or the cookiecutter template. **remove**

* ``morepath_wiki``. It's more or less of historical importance as it was one of the first workouts I gave of Morepath, but it does tie into the microframework battle presentation and shows how you can integrate Morepath with "alien" codebase (``storage.py``). It needs to be modernized (pep8, uppercase App class). **keep**

**Comment by henri-hulski** [comment]
@faassen
Maybe it's a good idea to move the samples you want to remove to faassen Github user. At least the ones which are still valuable like morepath_react. 

Like this the morepath organization is not cluttered with apps we don't maintain, but on the other hand they're still available for deeper exploring.

**Comment by faassen** [comment]
I'll move them over, as long as I won't have to maintain them. :)

**Comment by faassen** [comment]
Okay, I just moved morepath_react, morepath_hello and morepath_demo to my github account. While I was at it I also moved ``more.jsonld`` out because it was not a real used extension anyway.

**Comment by faassen** [comment]
I also renamed morepath_rest to morepath_rest_dump_load.

**Comment by henri-hulski** [comment]
So I have removed morepath_react, morepath_hello and morepath_demo from the task list.

@faassen 
So what you think about my idea to make now initial releases to freeze the status quo?

**Comment by faassen** [comment]
As an alternative to a release on github we could make a release on PyPI and make the PyPI page the canonical place we point to. We can include the README and changes there too automatically. We just need to make sure in MANIFEST.in that a full buildable source version of the project can be downloaded, and that there's a pointer back to the github project in the README.

This takes advantage from our tooling so that we can easily keep the published docs in sync.

**Comment by henri-hulski** [comment]
Ok, we can make releases on PyPI and Github in paralell if you think this makes sense.
I'm using now zest.releaser so there is no problem.
But I think this should be released under your PyPI user. Right?

> We can include the README and changes there too automatically

This is what I talk about inserting long_description in setup.py.

> We just need to make sure in MANIFEST.in that a full buildable source version of the project can be downloaded

zest.releaser checks if all code included in git is also included in MANIFEST.in.

**Comment by henri-hulski** [comment]
But I still think we can make the initial release Github only, more for historical reasons with the release date of the last commit done.
And after updating the samples, we can make a second release on Github and PyPI.

**Comment by faassen** [comment]
You can release them as your PyPI user and just add me and href as admins.

I'm not sure why we'd need a github release; it'd be a bit confusing as the subsequent releases would be on PyPI? I really don't see a reason to make a release -- the history is available on github anyway. Once we have releases on PyPI we can start updating them when needed --hopefully this is rare, but for now we just change stuff.

**Comment by henri-hulski** [comment]
> You can release them as your PyPI user and just add me and href as admins.

Not sure, where to do this. In setup.py or on pypi.python.org?

OK. So just starting updating the samples and when we finished we make a release on PyPI and Github in parallel to keep them in sync.

I think the best I use a separate branch on the morepath repos and make than a PR.
Like this I don't have to fork them but you can still review the changes.

That's ok for you?

**Comment by faassen** [comment]
You can give people admin rights in the PyPI admin UI, you should be able to see it for ``more.jwtauth``

I don't think we should be making github releases if we make PyPI releases anyway -- making a github release would just be extra manual work, and it's easy to forget and a missing release would confuse people.

I trust you to make the required changes in the trunk, especially when they're simple, but if you want to use a branch and then me a PR that's totally fine too.

**Comment by henri-hulski** [comment]
I think it would be a good idea to add some tests to the samples.
I will add this to the task list, but I'm not sure how far we go with it.

**Comment by faassen** [comment]
This one hasn't been updated for a while, anyone want to pitch in?

**Comment by henri-hulski** [comment]
I can release the ones I wrote tests for, if you're fine with it.

**Comment by faassen** [comment]
Sure, go ahead. Is the hold up documenting the applications? What exactly needs to be done?

**Comment by henri-hulski** [comment]
I think they are actually ready. I will just add a testenv for PY3.5 and release them.

**Comment by henri-hulski** [comment]
Released morepath_reactredux, morepath_sqlalchemy and morepath_rest_dump_load.

**Comment by henri-hulski** [comment]
> This one hasn't been updated for a while, anyone want to pitch in?

So we still need to write the tests for and release `morepath_batching` and `morepath_static`, if anyone wants to assist.

And write the documentation (#376).

**Comment by taschini** [comment]
I added the tests to ``morepath_static``.  I'll do the same for ``morepath_batching``.

**Comment by henri-hulski** [comment]
I think as #376 is closed we can also close this one. 
Or is there anything left to do?
